### PR TITLE
[Docs] Added go_api.md

### PR DESCRIPTION
**Pull request by spider0061**
Added go_api.md file.

References:
1. https://github.com/second-state/WasmEdge-go/blob/master/docs/go_api.md
2. https://www.secondstate.io/articles/extend-golang-app-with-webassembly-rust/
3. https://github.com/second-state/WasmEdge-go-examples
4. https://github.com/second-state/WasmEdge-go
5. https://github.com/second-state/WasmEdge-go-examples/tree/master/go_WasmAOT

### CODE REVIEW

**Comment by juntao** [comment]
Hi, This is a good start but I think we need more. Specifically, this should be a reference guide with explanations for all the public API functions the SDK has. Thanks.

**Comment by spider0061** [comment]
> Hi, This is a good start but I think we need more. Specifically, this should be a reference guide with explanations for all the public API functions the SDK has. Thanks.

Hello @juntao Where can I find all the available APIs? in examples?

**Comment by juntao** [comment]
In the SDK source code.

**Comment by spider0061** [comment]
> In the SDK source code.

This is the source code https://github.com/second-state/WasmEdge-go right?

**Comment by juntao** [comment]
Yes. That's the SDK. The reference documentation should cover most, if not all, public facing API classes, functions and their parameters. You can also investigate if you can generate these API docs automatically from golang source code comments using a tool like godoc.

**Comment by juntao** [comment]
The C API doc is an example as I mentioned. 

https://github.com/WasmEdge/WasmEdge/blob/master/docs/c_api.md

The Go SDK should be at very least this level of detail.

**Comment by spider0061** [comment]
> Yes. That's the SDK. The reference documentation should cover most, if not all, public facing API classes, functions and their parameters. You can also investigate if you can generate these API docs automatically from golang source code comments using a tool like godoc.

Got it thank you Sir

File: docs/go_api.md [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+In [this example](https://github.com/second-state/WasmEdge-go-examples/tree/master/go_BindgenFuncs), we will demonstrate how to call a few simple WebAssembly functions from a Golang app. The [functions](https://github.com/second-state/WasmEdge-go-examples/blob/master/go_BindgenFuncs/rust_bindgen_funcs/src/lib.rs) are written in Rust, and require complex call parameters and return values. The #[wasm_bindgen] macro is needed for the compiler tools to auto-generate the correct code to pass call parameters from Golang to WebAssembly.
+
+Note: The WebAssembly spec only supports a few simple data types out of the box. It [does not support](https://medium.com/wasm/strings-in-webassembly-wasm-57a05c1ea333) types such as string and array. In order to pass rich types in Golang to WebAssembly, the compiler needs to convert them to simple integers. For example, it converts a string into an integer memory address and an integer length. The wasm_bindgen tool, embedded in rustwasmc, does this conversion automatically.
+
+```go
```
Comment (q82419) [inline docs/go_api.md:85]: Should be `rust` ?

File: docs/go_api.md [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+```bash
+$ go version
+go version go1.16.5 linux/amd64
+```
+
+Developers must [install the WasmEdge shared library](https://github.com/WasmEdge/WasmEdge/blob/master/docs/install.md) with the same `WasmEdge-go` release version.
+
+```bash
+wget -qO- https://raw.githubusercontent.com/WasmEdge/WasmEdge/master/utils/install.sh | bash -s -- -v 0.9.0-rc.2
```
Comment (q82419) [inline docs/go_api.md:26]: Which version would you like to use in this documentation? `0.9.0-rc.2` or `0.8.2`?

**Review by q82419 — CHANGES_REQUESTED** [review]

**Comment by q82419** [comment]
> The C API doc is an example as I mentioned.
> 
> https://github.com/WasmEdge/WasmEdge/blob/master/docs/c_api.md
> 
> The Go SDK should be at very least this level of detail.

I think we can complete the details base on this starting and merge this PR first.
Only we know the tricks on wasmedge-go because of the lack of comments.

File: docs/go_api.md (hunk 2 above)
Comment (q82419) [reply docs/go_api.md:26]: For this document, I think use the `0.9.0-rc.2` and `WasmEdge-go v0.9.0-rc3` is better.

Comment (spider0061) [reply docs/go_api.md:26]: I have updated the document. Please have a look.

**Comment by spider0061** [comment]
> > The C API doc is an example as I mentioned.
> > https://github.com/WasmEdge/WasmEdge/blob/master/docs/c_api.md
> > The Go SDK should be at very least this level of detail.
> 
> I think we can complete the details base on this starting and merge this PR first. Only we know the tricks on wasmedge-go because of the lack of comments.

Sure

File: docs/go_api.md [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+```
+
+Noticed that the `TensorFlow` and `Image` extensions are only for the `Linux` platforms.
+
+Install the `WasmEdge-go` package and build in your Go project directory:
+
+```bash
+$ go get github.com/second-state/WasmEdge-go/wasmedge@v0.9.0-rc.2
```
Comment (q82419) [inline docs/go_api.md:42]: Should be `v0.9.0-rc3` here.

File: docs/go_api.md [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+    res, err = vm.ExecuteBindgen("sha3_digest", wasmedge.Bindgen_return_array, []byte("This is an important message"))
+    if err == nil {
+        fmt.Println("Run bindgen -- sha3_digest:", res.([]byte))
+    } 
+    res, err = vm.ExecuteBindgen("keccak_digest", wasmedge.Bindgen_return_array, []byte("This is an important message"))
+    if err == nil {
+        fmt.Println("Run bindgen -- keccak_digest:", res.([]byte))
+    } 
+
+    vm.Delete()
```
Comment (q82419) [inline docs/go_api.md:191]: For the newest API, we renamed the function name to `Release()`.

File: docs/go_api.md [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    conf.Delete()
```
Comment (q82419) [inline docs/go_api.md:192]: For the newest API, we renamed the function name to `Release()`.

File: docs/go_api.md [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+}
+```
+
+Next, let's build the Golang application with the WasmEdge Golang SDK.
+
+```bash
+$ go get -u github.com/second-state/WasmEdge-go/wasmedge
```
Comment (q82419) [inline docs/go_api.md:199]: It's better to add the version: `go get github.com/second-state/WasmEdge-go/wasmedge@v0.9.0-rc3`, or the `v0.8.2` will be get.

File: docs/go_api.md [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+        os.Args[1:],     /// The args
+        os.Environ(),    /// The envs
+        []string{".:."}, /// The mapping directories
+        []string{},      /// The preopens will be empty
+    )
+
+    /// Instantiate wasm. _start refers to the main() function
+    vm.RunWasmFile(os.Args[1], "_start")
+
+    vm.Delete()
```
Comment (q82419) [inline docs/go_api.md:309]: For the newest API, we renamed the function name to `Release()`.

File: docs/go_api.md [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+    conf.Delete()
```
Comment (q82419) [inline docs/go_api.md:310]: For the newest API, we renamed the function name to `Release()`.

File: docs/go_api.md [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+}
+```
+
+Next, let's build the Golang application with the WasmEdge Golang SDK.
+
+```bash
+$ go get -u github.com/second-state/WasmEdge-go
```
Comment (q82419) [inline docs/go_api.md:317]: It's better to add the version: `go get github.com/second-state/WasmEdge-go/wasmedge@v0.9.0-rc3`, or the `v0.8.2` will be get.

**Review by q82419 — CHANGES_REQUESTED** [review]

**Comment by spider0061** [comment]
@q82419 I have made the changes. Please have a look.

File: docs/go_api.md [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+```bash
+wget -qO- https://raw.githubusercontent.com/WasmEdge/WasmEdge/master/utils/install.sh | bash -s -- -v 0.9.0-rc.2
+```
+
+Noticed that the `TensorFlow` and `Image` extensions are only for the `Linux` platforms.
+
+Install the `WasmEdge-go` package and build in your Go project directory:
+
+```bash
+$ go get github.com/second-state/WasmEdge-go/wasmedge@vv0.9.0-rc3
```
Comment (q82419) [inline docs/go_api.md:42]: `v0.9.0-rc3`, not `vv0.9.0-rc3`.

**Review by q82419 — CHANGES_REQUESTED** [review]

**Comment by q82419** [comment]
Please squash your commits and add the `[Docs]` tag.

File: docs/go_api.md (hunk 10 above)
Comment (spider0061) [reply docs/go_api.md]: Done

**Comment by spider0061** [comment]
> Please squash your commits and add the `[Docs]` tag.

Squashed the commits and edited the message as well.

**Review by q82419 — APPROVED** [review]

**Pull request merged**